runelite-client/src/main/java/net/runelite/client/plugins/npcunaggroarea/NpcAggroAreaPlugin.java

```diff
@@ -0,0 +1,579 @@
+import net.runelite.api.events.GameTick;
+import net.runelite.api.events.NpcSpawned;
+import net.runelite.api.geometry.Geometry;
+import net.runelite.client.callback.ClientThread;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.game.ItemManager;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import net.runelite.client.ui.overlay.OverlayManager;
+import net.runelite.client.ui.overlay.infobox.InfoBoxManager;
+import net.runelite.client.ui.overlay.infobox.Timer;
+import net.runelite.client.util.WildcardMatcher;
+
+@Slf4j
+@PluginDescriptor(
+	name = "Unaggressive NPC timer",
+	description = "Highlights the unaggressive area of NPCs nearby and timer until it becomes active",
+	tags = {"highlight", "lines", "unaggro", "aggro", "aggressive", "npcs", "area", "timer", "slayer"}
+)
+public class NpcAggroAreaPlugin extends Plugin
+{
+	/*
+	How it works: The game remembers 2 tiles. When the player goes >10 steps
+	away from both tiles, the oldest one is moved to under the player and the
+	NPC aggression timer resets.
+	So to first figure out where the 2 tiles are, we wait until the player teleports
+	a long enough distance. At that point it's very likely that the player
+	moved out of the radius of both tiles, which resets one of them. The other
+	should reset shortly after as the player starts moving around.
+	*/
+
+	private final static int SAFE_AREA_RADIUS = 10;
+	private final static int UNKNOWN_AREA_RADIUS = SAFE_AREA_RADIUS * 2;
+	private final static int AGGRESSIVE_TIME_SECONDS = 600;
+	private final static Splitter NAME_SPLITTER = Splitter.on(CharMatcher.anyOf(",\n")).omitEmptyStrings().trimResults();
+
+	@Inject
+	private Client client;
+
+	@Inject
+	private NpcAggroAreaConfig config;
+
+	@Inject
+	private NpcAggroAreaOverlay overlay;
+
+	@Inject
+	private OverlayManager overlayManager;
+
+	@Inject
+	private ItemManager itemManager;
+
+	@Inject
+	private InfoBoxManager infoBoxManager;
+
+	@Inject
+	private ConfigManager configManager;
+
+	@Inject
+	private ClientThread clientThread;
+
+	@Getter
+	private WorldPoint[] safeCenters;
+
+	@Getter
+	private GeneralPath[] linesToDisplay;
+
+	@Getter
+	private Instant endTime;
+
+	@Getter
+	private boolean active;
+
+	private Timer currentTimer;
+	private WorldPoint lastPlayerLocation;
+	private int currentPlane;
+	private WorldPoint previousUnknownCenter;
+	private String username;
+	private boolean loggingIn;
+	private List<String> npcNamePatterns;
+
+	@Provides
+	NpcAggroAreaConfig provideConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(NpcAggroAreaConfig.class);
+	}
+
+	@Override
+	protected void startUp() throws Exception
+	{
+		overlayManager.add(overlay);
+		safeCenters = new WorldPoint[2];
+		linesToDisplay = new GeneralPath[Constants.MAX_Z];
+		npcNamePatterns = NAME_SPLITTER.splitToList(config.npcNamePatterns());
+
+		clientThread.invokeLater(() -> username = client.getUsername());
```

Comment:
this should be in GameStateChanged or even better, should not exist at all (just call client.getUsername() when needed)

Reply:
It is there because client.getUsername() is null after the player logs out (when gamestate changes to LOGIN_SCREEN). Though, if we're not using account specific config, the plugin doesn't need to know the username at all.